hnn_core/lfp.py

```diff
@@ -0,0 +1,407 @@
+        elif L > 0 and H < 0:  # electrode is "on top of" section
+            num = (np.sqrt(H ** 2 + R2) - H) * (L + np.sqrt(L ** 2 + R2))
+            denom = R2
+        else:  # electrode is "ahead of" section
+            num = np.sqrt(L ** 2 + R2) + L
+            denom = np.sqrt(H ** 2 + R2) + H  # == norm(b) + H
+
+        phi = np.log(num / denom) / norm_a
+
+    # [dis]: um; [sigma]: S / m
+    # [phi / sigma] = [1/dis] / [sigma] = 1 / [dis] x [sigma]
+    # [dis] x [sigma] = um x (S / m) = 1e-6 S
+    # transmembrane current returned by _ref_i_membrane_ is in [nA]
+    # ==> 1e-9 A x (1 / 1e-6 S) = 1e-3 V = mV
+    # ===> multiply by 1e3 to get uV
+    return 1000.0 * phi / (4.0 * np.pi * sigma)
+
+
+def _get_lfp_plot_data(times, data, tmin, tmax):
+    plot_tmin = times[0]
+    if tmin is not None:
+        plot_tmin = max(tmin, plot_tmin)
+    plot_tmax = times[-1]
+    if tmax is not None:
+        plot_tmax = min(tmax, plot_tmax)
+
+    times = np.array(times)
+    mask = np.logical_and(times >= plot_tmin, times < plot_tmax)
+    times = times[mask]
+
+    data = np.array(data)[mask]
+
+    return data, times
+
+
+class LFPArray:
+    """Local field potential (LFP) electrode array class.
+
+    Parameters
+    ----------
+    positions : tuple | list of tuple
+        The (x, y, z) coordinates (in um) of the LFP electrodes.
+    sigma : float
+        Extracellular conductivity, in S/m, of the assumed infinite,
+        homogeneous volume conductor that the cell and electrode are in.
+    method : str
+        Approximation to use. 'psa' (default) assigns all transmembrane
+        currents to the center point (0.5) (point source approximation).
+        'lsa' treats the section as a line source, and a single multiplier is
+        calculated for each section (line source approximation).
+
+    Attributes
+    ----------
+    voltages : list of float
+        The LFP voltages at the electrode positions (in uV).
+    times : list of float
+        The time points the LFP is sampled at (ms)
+    sfreq : float
+        Sampling rate of the LFP data (Hz).
+
+    Notes
+    -----
+    See Table 5 in http://jn.physiology.org/content/104/6/3388.long for
+    measured values of sigma in rat cortex (note units there are mS/cm)
+    """
+
+    def __init__(self, positions, sigma=0.3, method='psa'):
+        self.positions = positions
+        self.sigma = sigma
+        self.method = method
+        self.reset()
+
+    def __repr__(self):
+        class_name = self.__class__.__name__
+        msg = (f'{len(self.positions)} electrodes, sigma={self.sigma}, '
+               f'method={self.method}')
+        return f'<{class_name} | {msg}>'
+
+    def __len__(self):
+        return len(self.positions)
+
+    @property
+    def sfreq(self):
+        """Return the sampling rate of the LFP data."""
+        if len(self.times) > 1:
+            return 1000. / (self.times[1] - self.times[0])  # times in ms
```

Comment:
Maybe a test to make sure that all values in `np.diff(times)` are close to the `dt`?